server/src/middleware/merchant-auth.ts

```diff
@@ -17,25 +17,46 @@
 import {Request, Response, NextFunction} from 'express';
 import admin from 'firebase-admin';
 
+import {merchantService} from '../services';
+
 admin.initializeApp();
 
+const getVerifiedUid = async (firebaseIdToken: string): Promise<string> => {
+  const decodedFirebaseIdToken = await admin
+    .auth()
+    .verifyIdToken(firebaseIdToken);
+  return decodedFirebaseIdToken.uid;
+};
+
 const merchantAuth = async (
   req: Request,
   res: Response,
   next: NextFunction
 ) => {
   try {
-    if (!req.headers.authorization)
+    if (!req.headers.authorization) {
       throw new Error('Missing Authorization header.');
+    }
 
     const [_, firebaseIdToken] = req.headers.authorization.split(' ');
-    const decodedFirebaseIdToken = await admin
-      .auth()
-      .verifyIdToken(firebaseIdToken);
-    req.decoded = decodedFirebaseIdToken;
+    const verifiedFirebaseUid = await getVerifiedUid(firebaseIdToken);
+
+    let firebaseUid;
+    if (req.body.firebaseUid !== undefined) {
+      firebaseUid = req.body.firebaseUid;
+    } else if (req.body.merchantId !== undefined) {
+      const merchant = await merchantService.getMerchant(req.body.merchantId);
+      firebaseUid = merchant.firebaseUid;
+    }
```

Comment:
Can also check if firebaseUid is not defined (i.e. falling out of this if-elseIf block)?

Reply:
@xlai20 Yep, the check is in the next if block.

Reply:
You are referring to Line 52? That is not checking if firebaseUid is not defined right?
I'm thinking of (firebaseUid === undefined).

Reply:
@xlai20 Ohh, I see! Sorry I thought you meant something else.
My idea was that merchant-auth prevents the client from 'pretending' to be another merchant. But if the request body doesn't contain any reference to any merchant (hence `firebaseUid === undefined`), then they are not 'pretending' to be someone they are not. So, there is no error thrown here.
I was thinking that if the field (`firebaseUid` or `merchantId`) is needed for whatever operation, then the error would be thrown in the main function instead (i.e. service or storage layer).

Reply:
I got your idea that (firebaseUid==undefined) is not very likely to happen and will be caught by the caller on service/storage layer.
However, as a way of defensive programming, the function here alone should not completely trust the data coming from its caller and still have to do some error handling (or at least an assertion) of a case falling out of if-else-if block. Sometimes a software runs into bug in mysterious way and these small assertion/data-validation code helps a lot in debugging, especially when your code gets altered by the future you or someone else in a way that the current you cannot foresee.

Reply:
@xlai20 Based on our discussion in chat, the problem seems to be that the if-elseif block looks buggy. So I will refactor out this block of code to remove the use of if-elseif with no else.

Reply:
Nice, I'll approve first to not block your submission. Make sure to fix it based on our discussion.